### PR TITLE
feature/inference api demo impl

### PR DESCRIPTION
**Pull request by Superjomn**

### CODE REVIEW

File: paddle/contrib/inference/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+
+set(FLUID_CORE_MODULES proto_desc memory lod_tensor executor init)
```
Comment (luotao1) [inline paddle/contrib/inference/CMakeLists.txt]: 17行的set可以去掉。已经有这个变量了。

File: CMakeLists.txt [hunk 2]
```diff
@@ -60,6 +60,7 @@ option(WITH_DISTRIBUTE  "Compile with grpc distributed support"         OFF)
 option(USE_EIGEN_FOR_BLAS   "Use matrix multiplication in Eigen"        OFF)
 option(WITH_ARM_FP16    "Use half precision support on armv8.2-a cpu"   OFF)
 option(WITH_FAST_BUNDLE_TEST    "Bundle tests that can be run in a single process together to reduce launch overhead"   OFF)
+option(WITH_CONTRIB     "Compile the third-party contributation"        OFF)
```
Comment (luotao1) [inline CMakeLists.txt]: 需要修改一下https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/scripts/paddle_build.sh 的`cmake_gen`函数，加上`WITH_CONTRIB =ON`，不然ci测不了这部分代码。

Comment (Superjomn) [reply CMakeLists.txt]: 感觉应该要改teamcity

改成ON之后，CI可以过，但travis-android会挂

Comment (Superjomn) [reply CMakeLists.txt]: done

File: paddle/contrib/inference/paddle_inference_api.h [hunk 3]
```diff
@@ -12,49 +12,58 @@
...
+ */
+
 #pragma once
 
+#include <memory>
 #include <string>
 #include <vector>
 
-namespace paddle {
+struct PaddleTensor {
```
Comment (panyx0718) [inline paddle/contrib/inference/paddle_inference_api.h]: why no namespace?

File: paddle/contrib/inference/paddle_inference_api.h [hunk 4]
```diff
@@ -12,49 +12,58 @@
...
+  std::string name;  // variable name.
+  std::vector<int> shape;
+  std::vector<unsigned char> data;         // bytes of data.
```
Comment (panyx0718) [inline paddle/contrib/inference/paddle_inference_api.h]: this looks strange to me.
should it be 
```
std::unique_ptr<char> or perhaps just void* data, and user should delete it?
```

File: paddle/contrib/inference/paddle_inference_api.cc [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+/*
+ * Do not use this, just a demo indicating how to customize a Predictor.
+ */
+class DemoPredictor : public PaddlePredictor {
+public:
+  DemoPredictor() = default;
+
+  bool Init(const Config &config) override {
+    const DemoConfig &demo_config = *static_cast<const DemoConfig *>(&config);
```
Comment (panyx0718) [inline paddle/contrib/inference/paddle_inference_api.cc]: this looks strange to me....

File: paddle/contrib/inference/paddle_inference_api.h [hunk 6]
```diff
@@ -12,49 +12,58 @@
...
-           std::vector<std::vector<float>>* output_data);
-
-  // Clone a predictor that share the model weights.
-  Predictor* Clone();
+  virtual bool Run(const std::vector<PaddleTensor>& inputs,
+                   std::vector<PaddleTensor>* output_data) = 0;
+
+  // Clone a predictor that share the model weights, the Cloned predictor should
+  // be thread-safe.
+  virtual std::unique_ptr<PaddlePredictor> Clone() = 0;
```
Comment (panyx0718) [inline paddle/contrib/inference/paddle_inference_api.h]: If it's not required now, I suggest we leave it for future...Clone can easily be a source of bugs.

File: paddle/contrib/inference/paddle_inference_api.h [hunk 7]
```diff
@@ -12,49 +12,58 @@
...
+#include <memory>
 #include <string>
 #include <vector>
 
-namespace paddle {
+struct PaddleTensor {
+  std::string name;  // variable name.
+  std::vector<int> shape;
+  std::vector<unsigned char> data;         // bytes of data.
+  size_t type{typeid(float).hash_code()};  // hash of type
```
Comment (panyx0718) [inline paddle/contrib/inference/paddle_inference_api.h]: not sure typeid is a good idea. What are the typeid for float16 and int8, and mabe complex32 in the future...

Maybe enum?

File: paddle/contrib/inference/paddle_inference_api.h [hunk 8]
```diff
@@ -12,49 +12,58 @@
...
+  // Clone a predictor that share the model weights, the Cloned predictor should
+  // be thread-safe.
+  virtual std::unique_ptr<PaddlePredictor> Clone() = 0;
 
   // Destroy the Predictor.
-  ~Predictor();
+  virtual ~PaddlePredictor() {}
 
-  struct Attr {
+  friend std::unique_ptr<PaddlePredictor> CreatePaddlePredictor(
```
Comment (panyx0718) [inline paddle/contrib/inference/paddle_inference_api.h]: why friend?

File: paddle/contrib/inference/paddle_inference_api.h [hunk 9]
```diff
@@ -64,6 +73,18 @@ class Predictor {
...
 };
 
-}  // namespace paddle
+// A factory to help create difference predictor.
+std::unique_ptr<PaddlePredictor> CreatePaddlePredictor(
+    const PaddlePredictor::Config& config);
+
+// Do not use this, just a demo indicating how to customize a config for a
+// specific predictor.
+struct DemoConfig : public PaddlePredictor::Config {
```
Comment (panyx0718) [inline paddle/contrib/inference/paddle_inference_api.h]: Let's be more serious. DemoXXX shouldn't show up in api header file... move it to test.cc?

File: paddle/contrib/inference/paddle_inference_api.h (hunk 3 above)
Comment (Superjomn) [reply paddle/contrib/inference/paddle_inference_api.h]: the namespace will make the symbol name more complex if user use `dlsym` to load the APIs from a dynamic library.

Just add `paddle` prefix to each API, and the most important method `CreatePaddlePredictor` can load easily by `dlsym`.

File: paddle/contrib/inference/paddle_inference_api.h (hunk 4 above)
Comment (Superjomn) [reply paddle/contrib/inference/paddle_inference_api.h]: The `data` is just a vector of bytes, the memory can be reused and resized, compared to `unique_ptr` memory.

File: paddle/contrib/inference/paddle_inference_api.h (hunk 6 above)
Comment (Superjomn) [reply paddle/contrib/inference/paddle_inference_api.h]: Currently, just return `nullptr`, this API is required by the image team.

File: paddle/contrib/inference/paddle_inference_api.cc (hunk 5 above)
Comment (Superjomn) [reply paddle/contrib/inference/paddle_inference_api.cc]: I will change it tomorrow, seems a template specification more natural here.

File: paddle/contrib/inference/paddle_inference_api.h (hunk 4 above)
Comment (panyx0718) [reply paddle/contrib/inference/paddle_inference_api.h]: not sure user would want to to resize memory. I rarely see putting a blob of memory in vector. if not used correctly (e.g. push_back), the memory actually used can be unpredictable.

File: paddle/contrib/inference/paddle_inference_api.h (hunk 6 above)
Comment (panyx0718) [reply paddle/contrib/inference/paddle_inference_api.h]: I see.

File: paddle/contrib/inference/paddle_inference_api.h (hunk 4 above)
Comment (Superjomn) [reply paddle/contrib/inference/paddle_inference_api.h]: The memory will be resized for different batch.

File: paddle/contrib/inference/paddle_inference_api.h (hunk 7 above)
Comment (Superjomn) [reply paddle/contrib/inference/paddle_inference_api.h]: Yes, the `typeid` is unique for a different type, it is a global id compared to an enum.

 > size_t type{typeid(float).hash_code()};

this is a default value, users can set with a `typeid` of any type.

**Pull request merged**